File: ios/StackedTextReader.swift
//
//  StackedTextReader.swift
//  react-native-vision-camera-ml-kit
//

import Foundation
import UIKit
import CoreGraphics
import MLKitVision
import MLKitTextRecognition

final class StackedTextReader {

    struct GlyphReading {
        let text: String
        let box: CGRect
    }

    struct StackedBlock {
        let text: String
        let bounds: CGRect
        let glyphs: [GlyphReading]
    }

    struct Prepared {
        let strip: UIImage
        let bounds: CGRect
        let glyphBoxes: [CGRect]
        let widths: [Int]
    }

    static let frameDetectionLongSide = 1920
    static let staticDetectionLongSide = 1200
    static let maxColumnsToRead = 6
    private static let tileHeight = 96
    private static let tileGap = 24
    private static let minTileWidth = 8
    private static let glyphPadding = 0.20
    private static let colourSamples = 24
    private static let rotationSweep = [0, 90, 270, 180]

    private let workspace = StackedTextDetector.Workspace()

    func read(_ image: UIImage, recognizer: TextRecognizer) -> [StackedBlock] {
        guard let source = image.cgImage else { return [] }
        let sourceWidth = source.width
        let sourceHeight = source.height
        guard sourceWidth >= 2, sourceHeight >= 2 else { return [] }

        let started = Date()
        let scale = min(1.0, Double(Self.staticDetectionLongSide) / Double(max(sourceWidth, sourceHeight)))
        let detectionWidth = max(1, Int((Double(sourceWidth) * scale).rounded()))
        let detectionHeight = max(1, Int((Double(sourceHeight) * scale).rounded()))

        guard let rgba = Self.rgbaBuffer(source, width: detectionWidth, height: detectionHeight) else {
            Logger.warn("Could not read pixels for stacked detection")
            return []
        }

        let columns = StackedTextDetector.detect(rgba: rgba, width: detectionWidth, height: detectionHeight, workspace: workspace)
        Logger.performance("stacked.detect", durationMs: Int64(Date().timeIntervalSince(started) * 1000))
        Logger.debug("stacked.columns=\(columns.count) \(columns.map { $0.glyphs.count }) in \(detectionWidth)x\(detectionHeight)")

        let prepared = prepare(columns, inverse: 1.0 / scale, source: source, rotationDegrees: 0, uprightWidth: sourceWidth, uprightHeight: sourceHeight)
        return recognize(prepared, recognizer: recognizer)
    }

    func readWithRotationFallback(_ image: UIImage, recognizer: TextRecognizer) -> [StackedBlock] {
        for degrees in Self.rotationSweep {
            let rotated = Self.rotatedCopy(image, degrees: degrees)
            let blocks = read(rotated, recognizer: recognizer)
            if blocks.isEmpty { continue }
            if degrees == 0 { return blocks }
            let size = rotated.size
            return blocks.map { Self.toSourceCoordinates($0, degrees: degrees, rotatedSize: size) }
        }
        return []
    }

    func prepare(luma: ImageUtils.LumaPlane, source: CGImage, rotationDegrees: Int) -> [Prepared] {
        let started = Date()
        let columns = StackedTextDetector.detect(luma: luma.bytes, width: luma.width, height: luma.height, workspace: workspace)
        Logger.performance("stacked.detect", durationMs: Int64(Date().timeIntervalSince(started) * 1000))
        Logger.debug("stacked.columns=\(columns.count) \(columns.map { $0.glyphs.count }) in \(luma.width)x\(luma.height), passes=\(workspace.passesRun)")

        let upright = rotationDegrees % 180 == 0
        let uprightWidth = upright ? source.width : source.height
        let uprightHeight = upright ? source.height : source.width
        return prepare(columns, inverse: Double(luma.step), source: source, rotationDegrees: rotationDegrees, uprightWidth: uprightWidth, uprightHeight: uprightHeight)
    }

    func recognize(_ prepared: [Prepared], recognizer: TextRecognizer) -> [StackedBlock] {
        prepared.compactMap { readStrip($0, recognizer: recognizer) }
    }

    private func prepare(
        _ columns: [StackedTextDetector.Column],
        inverse: Double,
        source: CGImage,
        rotationDegrees: Int,
        uprightWidth: Int,
        uprightHeight: Int
    ) -> [Prepared] {
        if columns.isEmpty { return [] }
        let started = Date()
        let ranked = columns
            .map { ($0, StackedTextDetector.refine($0)) }
            .sorted { $0.1.score < $1.1.score }
            .prefix(Self.maxColumnsToRead)

        var prepared = [Prepared]()
        for (column, refined) in ranked {
            let glyphBoxes = refined.glyphs.map { Self.toSourceRect($0, inverse: inverse, width: uprightWidth, height: uprightHeight) }
            let tiles = glyphBoxes.indices.map {
                Self.padded(glyphBoxes, index: $0, boxed: refined.boxed[$0], width: uprightWidth, height: uprightHeight)
            }
            let widths = tiles.map { tile -> Int in
                max(Self.minTileWidth, Int((tile.width * CGFloat(Self.tileHeight) / tile.height).rounded()))
            }
            let region = tiles.dropFirst().reduce(tiles[0]) { $0.union($1) }.integral

            guard let crop = Self.uprightCrop(source, region: region, rotationDegrees: rotationDegrees, uprightWidth: uprightWidth, uprightHeight: uprightHeight) else {
                continue
            }
            let strip = Self.buildStrip(crop: crop, tiles: tiles, offset: region.origin, widths: widths, background: Self.medianColour(crop))
            prepared.append(Prepared(
                strip: strip,
                bounds: Self.toSourceRect(column.bounds, inverse: inverse, width: uprightWidth, height: uprightHeight),
                glyphBoxes: glyphBoxes,
                widths: widths
            ))
        }
        Logger.performance("stacked.strip.build", durationMs: Int64(Date().timeIntervalSince(started) * 1000))
        return prepared
    }

    private func readStrip(_ prepared: Prepared, recognizer: TextRecognizer) -> StackedBlock? {
        let started = Date()
        let visionImage = VisionImage(image: prepared.strip)
        visionImage.orientation = .up

        let text: Text
        do {
            text = try recognizer.results(in: visionImage)
        } catch {
            Logger.error("Stacked strip recognition failed: \(error.localizedDescription)")
            return nil
        }
        Logger.performance("stacked.strip.mlkit", durationMs: Int64(Date().timeIntervalSince(started) * 1000))

        let lines = text.blocks.flatMap { $0.lines }.sorted { $0.frame.minX < $1.frame.minX }
        let reading = lines.map { $0.text }.joined(separator: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        Logger.debug("Stacked strip read: '\(reading)'")
        if reading.isEmpty { return nil }

        return StackedBlock(
            text: reading,
            bounds: prepared.bounds,
            glyphs: Self.assignGlyphText(text, glyphBoxes: prepared.glyphBoxes, widths: prepared.widths)
        )
    }

    private static func uprightCrop(
        _ source: CGImage,
        region: CGRect,
        rotationDegrees: Int,
        uprightWidth: Int,
        uprightHeight: Int
    ) -> CGImage? {
        let rotatedSize = CGSize(width: uprightWidth, height: uprightHeight)
        let inSource = toSourceCoordinates(region, degrees: rotationDegrees, rotatedSize: rotatedSize)
        guard let cropped = source.cropping(to: inSource) else { return nil }
        if rotationDegrees % 360 == 0 { return cropped }
        return rotatedCopy(UIImage(cgImage: cropped), degrees: rotationDegrees).cgImage
    }

    private static func buildStrip(
        crop: CGImage,
        tiles: [CGRect],
        offset: CGPoint,
        widths: [Int],
        background: UIColor
    ) -> UIImage {
        let size = CGSize(
            width: widths.reduce(0, +) + tileGap * (tiles.count + 1),
            height: tileHeight + tileGap * 2
        )
        let format = UIGraphicsImageRendererFormat.default()
        format.scale = 1
        format.opaque = true

        return UIGraphicsImageRenderer(size: size, format: format).image { context in
            background.setFill()
            context.fill(CGRect(origin: .zero, size: size))

            var x = CGFloat(tileGap)
            for index in tiles.indices {
                let tile = tiles[index].offsetBy(dx: -offset.x, dy: -offset.y)
                guard let cropped = crop.cropping(to: tile) else {
                    x += CGFloat(widths[index] + tileGap)
                    continue
                }
                let destination = CGRect(
                    x: x,
                    y: CGFloat(tileGap),
                    width: CGFloat(widths[index]),
                    height: CGFloat(tileHeight)
                )
                UIImage(cgImage: cropped).draw(in: destination)
                x += CGFloat(widths[index] + tileGap)
            }
        }
    }

    private static func assignGlyphText(
        _ text: Text,
        glyphBoxes: [CGRect],
        widths: [Int]
    ) -> [GlyphReading] {
        var readings = [String](repeating: "", count: glyphBoxes.count)
        var starts = [CGFloat]()
        var x = CGFloat(tileGap)
        for width in widths {
            starts.append(x)
            x += CGFloat(width + tileGap)
        }

        for block in text.blocks {
            for line in block.lines {
                for element in line.elements {
                    let characters = Array(element.text)
                    if characters.isEmpty { continue }
                    let frame = element.frame
                    let step = frame.width / CGFloat(characters.count)
                    for (offset, character) in characters.enumerated() {
                        let centre = frame.minX + step * (CGFloat(offset) + 0.5)
                        guard let tile = starts.lastIndex(where: { centre >= $0 }) else { continue }
                        if centre > starts[tile] + CGFloat(widths[tile]) { continue }
                        readings[tile].append(character)
                    }
                }
            }
        }

        return glyphBoxes.enumerated().map { GlyphReading(text: readings[$0.offset], box: $0.element) }
    }

    private static func rgbaBuffer(_ source: CGImage, width: Int, height: Int) -> [UInt8]? {
        var buffer = [UInt8](repeating: 0, count: width * height * 4)
        let bitmapInfo = CGImageAlphaInfo.premultipliedLast.rawValue
        let drawn: Bool = buffer.withUnsafeMutableBytes { raw -> Bool in
            guard let base = raw.baseAddress,
                  let context = CGContext(
                    data: base,
                    width: width,
                    height: height,
                    bitsPerComponent: 8,
                    bytesPerRow: width * 4,
                    space: CGColorSpaceCreateDeviceRGB(),
                    bitmapInfo: bitmapInfo
                  ) else { return false }
            context.interpolationQuality = .high
            context.draw(source, in: CGRect(x: 0, y: 0, width: width, height: height))
            return true
        }
        return drawn ? buffer : nil
    }

    private static func medianColour(_ crop: CGImage) -> UIColor {
        let width = crop.width
        let height = crop.height
        guard width > 0, height > 0, let pixels = rgbaBuffer(crop, width: width, height: height) else { return .black }

        let stepX = max(1, width / colourSamples)
        let stepY = max(1, height / colourSamples)
        var reds = [Int](), greens = [Int](), blues = [Int]()
        var y = 0
        while y < height {
            var x = 0
            while x < width {
                let base = (y * width + x) * 4
                reds.append(Int(pixels[base]))
                greens.append(Int(pixels[base + 1]))
                blues.append(Int(pixels[base + 2]))
                x += stepX
            }
            y += stepY
        }
        guard !reds.isEmpty else { return .black }
        let mid = reds.count / 2
        return UIColor(
            red: CGFloat(reds.sorted()[mid]) / 255.0,
            green: CGFloat(greens.sorted()[mid]) / 255.0,
            blue: CGFloat(blues.sorted()[mid]) / 255.0,
            alpha: 1
        )
    }

    private static func toSourceRect(
        _ box: StackedTextDetector.Box,
        inverse: Double,
        width: Int,
        height: Int
    ) -> CGRect {
        let left = min(max(0, (Double(box.left) * inverse).rounded()), Double(width - 1))
        let top = min(max(0, (Double(box.top) * inverse).rounded()), Double(height - 1))
        let right = min(max(left + 1, (Double(box.right) * inverse).rounded()), Double(width))
        let bottom = min(max(top + 1, (Double(box.bottom) * inverse).rounded()), Double(height))
        return CGRect(x: left, y: top, width: right - left, height: bottom - top)
    }

    private static func padded(_ all: [CGRect], index: Int, boxed: Bool, width: Int, height: Int) -> CGRect {
        let box = all[index]
        let pad = boxed ? 0 : (box.height * CGFloat(glyphPadding)).rounded()
        let gapUp = index > 0 ? box.minY - all[index - 1].maxY : pad * 2
        let gapDown = index + 1 < all.count ? all[index + 1].minY - box.maxY : pad * 2
        let padY = min(pad, max(0, gapUp / 2), max(0, gapDown / 2)).rounded(.down)
        let left = min(max(0, box.minX - pad), CGFloat(width - 1))
        let top = min(max(0, box.minY - padY), CGFloat(height - 1))
        let right = min(max(left + 1, box.maxX + pad), CGFloat(width))
        let bottom = min(max(top + 1, box.maxY + padY), CGFloat(height))
        return CGRect(x: left, y: top, width: right - left, height: bottom - top)
    }

    static func rotatedCopy(_ image: UIImage, degrees: Int) -> UIImage {
        let normalized = ((degrees % 360) + 360) % 360
        guard normalized != 0, let source = image.cgImage else { return image }

        let width = CGFloat(source.width)
        let height = CGFloat(source.height)
        let size = normalized == 180
            ? CGSize(width: width, height: height)
            : CGSize(width: height, height: width)

        let format = UIGraphicsImageRendererFormat.default()
        format.scale = 1
        format.opaque = true

        return UIGraphicsImageRenderer(size: size, format: format).image { context in
            let cgContext = context.cgContext
            cgContext.translateBy(x: size.width / 2, y: size.height / 2)
            cgContext.rotate(by: CGFloat(normalized) * .pi / 180)
            cgContext.translateBy(x: -width / 2, y: -height / 2)
            UIImage(cgImage: source).draw(in: CGRect(x: 0, y: 0, width: width, height: height))
        }
    }

    static func toSourceCoordinates(_ block: StackedBlock, degrees: Int, rotatedSize: CGSize) -> StackedBlock {
        if ((degrees % 360) + 360) % 360 == 0 { return block }
        return StackedBlock(
            text: block.text,
            bounds: toSourceCoordinates(block.bounds, degrees: degrees, rotatedSize: rotatedSize),
            glyphs: block.glyphs.map {
                GlyphReading(
                    text: $0.text,
                    box: toSourceCoordinates($0.box, degrees: degrees, rotatedSize: rotatedSize)
                )
            }
        )
    }

    static func toSourceCoordinates(_ rect: CGRect, degrees: Int, rotatedSize: CGSize) -> CGRect {
        switch ((degrees % 360) + 360) % 360 {
        case 90:
            return CGRect(
                x: rect.minY,
                y: rotatedSize.width - rect.maxX,
                width: rect.height,
                height: rect.width
            )
        case 180:
            return CGRect(
                x: rotatedSize.width - rect.maxX,
                y: rotatedSize.height - rect.maxY,
                width: rect.width,
                height: rect.height
            )
        case 270:
            return CGRect(
                x: rotatedSize.height - rect.maxY,
                y: rect.minX,
                width: rect.height,
                height: rect.width
            )
        default:
            return rect
        }
    }
}

File: android/src/main/java/com/rnvisioncameramlkit/stacked/StackedTextReader.kt
package com.rnvisioncameramlkit.stacked

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.Rect
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.Text
import com.google.mlkit.vision.text.TextRecognizer
import com.rnvisioncameramlkit.utils.ImageUtils
import com.rnvisioncameramlkit.utils.Logger
import java.util.concurrent.TimeUnit
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

class StackedTextReader {

    data class GlyphReading(val text: String, val box: Rect)

    data class StackedBlock(val text: String, val bounds: Rect, val glyphs: List<GlyphReading>)

    class Prepared(val strip: Bitmap, val bounds: Rect, val glyphBoxes: List<Rect>, val widths: IntArray)

    private val workspace = StackedTextDetector.Workspace()
    private var pixels = IntArray(0)
    private val reds = IntArray(COLOUR_SAMPLES * COLOUR_SAMPLES)
    private val greens = IntArray(COLOUR_SAMPLES * COLOUR_SAMPLES)
    private val blues = IntArray(COLOUR_SAMPLES * COLOUR_SAMPLES)

    fun read(source: Bitmap, recognizer: TextRecognizer, timeoutSeconds: Long): List<StackedBlock> {
        val started = System.currentTimeMillis()
        val scale = min(1f, STATIC_DETECTION_LONG_SIDE.toFloat() / max(source.width, source.height))
        val detectionWidth = max(1, (source.width * scale).roundToInt())
        val detectionHeight = max(1, (source.height * scale).roundToInt())

        val scaled = if (detectionWidth == source.width && detectionHeight == source.height) source
        else Bitmap.createScaledBitmap(source, detectionWidth, detectionHeight, true)

        val columns: List<StackedTextDetector.Column>
        try {
            val count = scaled.width * scaled.height
            if (pixels.size < count) pixels = IntArray(count)
            scaled.getPixels(pixels, 0, scaled.width, 0, 0, scaled.width, scaled.height)
            columns = StackedTextDetector.detect(pixels, scaled.width, scaled.height, workspace)
        } finally {
            if (scaled !== source) scaled.recycle()
        }
        Logger.performance("stacked.detect", System.currentTimeMillis() - started)
        Logger.debug("stacked.columns=${columns.size} ${columns.map { it.glyphs.size }} in ${detectionWidth}x${detectionHeight}")

        val prepared = prepare(columns, 1f / scale, source, 0, source.width, source.height)
        return recognize(prepared, recognizer, timeoutSeconds)
    }

    fun readWithRotationFallback(
        source: Bitmap,
        recognizer: TextRecognizer,
        timeoutSeconds: Long
    ): List<StackedBlock> {
        for (degrees in ROTATION_SWEEP) {
            val rotated = rotatedCopy(source, degrees)
            try {
                val blocks = read(rotated, recognizer, timeoutSeconds)
                if (blocks.isEmpty()) continue
                if (degrees == 0) return blocks
                return blocks.map { block ->
                    StackedBlock(
                        block.text,
                        toSourceCoordinates(block.bounds, degrees, rotated.width, rotated.height),
                        block.glyphs.map {
                            GlyphReading(
                                it.text,
                                toSourceCoordinates(it.box, degrees, rotated.width, rotated.height)
                            )
                        }
                    )
                }
            } finally {
                if (rotated !== source) rotated.recycle()
            }
        }
        return emptyList()
    }

    fun prepare(luma: ImageUtils.LumaPlane, source: Bitmap, rotationDegrees: Int): List<Prepared> {
        val started = System.currentTimeMillis()
        val columns = StackedTextDetector.detect(luma.bytes, luma.width, luma.height, workspace)
        Logger.performance("stacked.detect", System.currentTimeMillis() - started)
        Logger.debug("stacked.columns=${columns.size} ${columns.map { it.glyphs.size }} in ${luma.width}x${luma.height}, passes=${workspace.passesRun}")

        val upright = rotationDegrees % 180 == 0
        val uprightWidth = if (upright) source.width else source.height
        val uprightHeight = if (upright) source.height else source.width
        return prepare(columns, luma.step.toFloat(), source, rotationDegrees, uprightWidth, uprightHeight)
    }

    fun recognize(prepared: List<Prepared>, recognizer: TextRecognizer, timeoutSeconds: Long): List<StackedBlock> {
        try {
            return prepared.mapNotNull { readStrip(it, recognizer, timeoutSeconds) }
        } finally {
            prepared.forEach { it.strip.recycle() }
        }
    }

    private fun prepare(
        columns: List<StackedTextDetector.Column>,
        inverse: Float,
        source: Bitmap,
        rotationDegrees: Int,
        uprightWidth: Int,
        uprightHeight: Int
    ): List<Prepared> {
        if (columns.isEmpty()) return emptyList()
        val started = System.currentTimeMillis()
        val ranked = columns
            .map { it to StackedTextDetector.refine(it) }
            .sortedBy { it.second.score }
            .take(MAX_COLUMNS_TO_READ)

        val prepared = ArrayList<Prepared>(ranked.size)
        try {
            for ((column, refined) in ranked) {
                val glyphBoxes = refined.glyphs.map { it.toSourceRect(inverse, uprightWidth, uprightHeight) }
                val tiles = glyphBoxes.indices.map { padded(glyphBoxes, it, refined.boxed[it], uprightWidth, uprightHeight) }
                val widths = IntArray(tiles.size) {
                    max(MIN_TILE_WIDTH, (tiles[it].width() * TILE_HEIGHT.toFloat() / tiles[it].height()).roundToInt())
                }
                val region = Rect(tiles.first())
                for (i in 1 until tiles.size) region.union(tiles[i])

                val crop = uprightCrop(source, region, rotationDegrees, uprightWidth, uprightHeight)
                try {
                    prepared += Prepared(
                        buildStrip(crop, tiles, region.left, region.top, widths),
                        column.bounds.toSourceRect(inverse, uprightWidth, uprightHeight),
                        glyphBoxes,
                        widths
                    )
                } finally {
                    if (crop !== source) crop.recycle()
                }
            }
        } catch (e: Exception) {
            prepared.forEach { it.strip.recycle() }
            throw e
        }
        Logger.performance("stacked.strip.build", System.currentTimeMillis() - started)
        return prepared
    }

    private fun readStrip(prepared: Prepared, recognizer: TextRecognizer, timeoutSeconds: Long): StackedBlock? {
        val started = System.currentTimeMillis()
        val text: Text
        try {
            text = Tasks.await(
                recognizer.process(InputImage.fromBitmap(prepared.strip, 0)),
                timeoutSeconds,
                TimeUnit.SECONDS
            )
        } catch (e: Exception) {
            Logger.error("Stacked strip recognition failed", e)
            return null
        }
        Logger.performance("stacked.strip.mlkit", System.currentTimeMillis() - started)

        val lines = text.textBlocks.flatMap { it.lines }.sortedBy { it.boundingBox?.left ?: 0 }
        val reading = lines.joinToString(" ") { it.text }.trim()
        Logger.debug("Stacked strip read: '$reading'")
        if (reading.isEmpty()) return null

        return StackedBlock(reading, prepared.bounds, assignGlyphText(text, prepared.glyphBoxes, prepared.widths))
    }

    private fun uprightCrop(
        source: Bitmap,
        region: Rect,
        rotationDegrees: Int,
        uprightWidth: Int,
        uprightHeight: Int
    ): Bitmap {
        val inSource = toSourceCoordinates(region, rotationDegrees, uprightWidth, uprightHeight)
        val matrix = if (rotationDegrees % 360 == 0) null else Matrix().apply { postRotate(rotationDegrees.toFloat()) }
        return Bitmap.createBitmap(source, inSource.left, inSource.top, inSource.width(), inSource.height(), matrix, true)
    }

    private fun buildStrip(crop: Bitmap, tiles: List<Rect>, offsetX: Int, offsetY: Int, widths: IntArray): Bitmap {
        val strip = Bitmap.createBitmap(
            widths.sum() + TILE_GAP * (tiles.size + 1),
            TILE_HEIGHT + TILE_GAP * 2,
            Bitmap.Config.ARGB_8888
        )
        val canvas = Canvas(strip)
        canvas.drawColor(medianColour(crop))
        val paint = Paint(Paint.FILTER_BITMAP_FLAG)
        val tile = Rect()
        var x = TILE_GAP
        for (i in tiles.indices) {
            tile.set(tiles[i])
            tile.offset(-offsetX, -offsetY)
            canvas.drawBitmap(crop, tile, Rect(x, TILE_GAP, x + widths[i], TILE_GAP + TILE_HEIGHT), paint)
            x += widths[i] + TILE_GAP
        }
        return strip
    }

    private fun assignGlyphText(text: Text, glyphBoxes: List<Rect>, widths: IntArray): List<GlyphReading> {
        val readings = Array(glyphBoxes.size) { StringBuilder() }
        val starts = IntArray(widths.size)
        var x = TILE_GAP
        for (i in widths.indices) {
            starts[i] = x
            x += widths[i] + TILE_GAP
        }

        for (block in text.textBlocks) {
            for (line in block.lines) {
                for (element in line.elements) {
                    for (symbol in element.symbols) {
                        val box = symbol.boundingBox ?: continue
                        val centre = box.exactCenterX()
                        val tile = starts.indices.lastOrNull { centre >= starts[it] } ?: continue
                        if (centre > starts[tile] + widths[tile]) continue
                        readings[tile].append(symbol.text)
                    }
                }
            }
        }

        return glyphBoxes.mapIndexed { i, box -> GlyphReading(readings[i].toString(), box) }
    }

    private fun medianColour(crop: Bitmap): Int {
        val stepY = max(1, crop.height / COLOUR_SAMPLES)
        val stepX = max(1, crop.width / COLOUR_SAMPLES)
        var n = 0
        var y = 0
        while (y < crop.height && n < reds.size) {
            var x = 0
            while (x < crop.width && n < reds.size) {
                val colour = crop.getPixel(x, y)
                reds[n] = Color.red(colour)
                greens[n] = Color.green(colour)
                blues[n] = Color.blue(colour)
                n++
                x += stepX
            }
            y += stepY
        }
        if (n == 0) return Color.BLACK
        reds.sort(0, n)
        greens.sort(0, n)
        blues.sort(0, n)
        return Color.rgb(reds[n / 2], greens[n / 2], blues[n / 2])
    }

    private fun padded(all: List<Rect>, index: Int, boxed: Boolean, width: Int, height: Int): Rect {
        val box = all[index]
        val pad = if (boxed) 0 else (box.height() * GLYPH_PADDING).roundToInt()
        val gapUp = if (index > 0) box.top - all[index - 1].bottom else pad * 2
        val gapDown = if (index + 1 < all.size) all[index + 1].top - box.bottom else pad * 2
        val padY = min(pad, min(max(0, gapUp / 2), max(0, gapDown / 2)))
        val left = (box.left - pad).coerceIn(0, width - 1)
        val top = (box.top - padY).coerceIn(0, height - 1)
        return Rect(
            left,
            top,
            (box.right + pad).coerceIn(left + 1, width),
            (box.bottom + padY).coerceIn(top + 1, height)
        )
    }

    private fun StackedTextDetector.Box.toSourceRect(inverse: Float, width: Int, height: Int): Rect {
        val left = (left * inverse).roundToInt().coerceIn(0, width - 1)
        val top = (top * inverse).roundToInt().coerceIn(0, height - 1)
        return Rect(
            left,
            top,
            (right * inverse).roundToInt().coerceIn(left + 1, width),
            (bottom * inverse).roundToInt().coerceIn(top + 1, height)
        )
    }

    companion object {
        const val FRAME_DETECTION_LONG_SIDE = 1920
        const val STATIC_DETECTION_LONG_SIDE = 1200
        const val MAX_COLUMNS_TO_READ = 6
        private const val TILE_HEIGHT = 96
        private const val TILE_GAP = 24
        private const val MIN_TILE_WIDTH = 8
        private const val GLYPH_PADDING = 0.20f
        private const val COLOUR_SAMPLES = 24
        private val ROTATION_SWEEP = intArrayOf(0, 90, 270, 180)

        fun rotatedCopy(source: Bitmap, degrees: Int): Bitmap {
            if (degrees % 360 == 0) return source
            val matrix = Matrix().apply { postRotate(degrees.toFloat()) }
            return Bitmap.createBitmap(source, 0, 0, source.width, source.height, matrix, true)
        }

        fun toSourceCoordinates(rect: Rect, degrees: Int, rotatedWidth: Int, rotatedHeight: Int): Rect =
            when ((degrees % 360 + 360) % 360) {
                90 -> Rect(rect.top, rotatedWidth - rect.right, rect.bottom, rotatedWidth - rect.left)
                180 -> Rect(
                    rotatedWidth - rect.right,
                    rotatedHeight - rect.bottom,
                    rotatedWidth - rect.left,
                    rotatedHeight - rect.top
                )
                270 -> Rect(rotatedHeight - rect.bottom, rect.left, rotatedHeight - rect.top, rect.right)
                else -> Rect(rect)
            }
    }
}
